src/transformers/models/molmo/modular_molmo.py

```diff
@@ -0,0 +1,2289 @@
+            )
+            valid_batch_indices = batch_indices[valid_crops_flat]
+            # now all valid crops together
+            image_features = self.get_image_features(
+                pixel_values=all_pixel_values.unsqueeze(1),
+                image_masks=all_image_masks.unsqueeze(1),
+                vision_feature_layers=vision_feature_layers,
+                vision_feature_select_strategy=vision_feature_select_strategy,
+            )  # this returns [total_valid_crops, num_image_tokens, hidden_size]
+
+            image_features_flat = image_features.view(-1, hidden_size)
+            image_token_indices_flat = all_image_token_indices.view(-1)
+
+            valid_indices_mask = image_token_indices_flat != -100
+            image_token_indices_flat[valid_indices_mask] += 1  # adjustment, TODO is this still needed
+
+            valid_batch_indices_expanded = (
+                valid_batch_indices.unsqueeze(1).expand(-1, all_image_token_indices.size(-1)).reshape(-1)
+            )
+
+            valid_positions = image_token_indices_flat >= 0
+            valid_indices = image_token_indices_flat[valid_positions].long()
+            valid_features = image_features_flat[valid_positions]
+            valid_batch_indices = valid_batch_indices_expanded[valid_positions].long()
+
+            flat_indices = valid_batch_indices * seq_len + valid_indices
+            inputs_embeds_flat = inputs_embeds.view(-1, hidden_size)
+
+            inputs_embeds_flat.index_add_(0, flat_indices, valid_features.to(inputs_embeds_flat.device))
+            inputs_embeds = inputs_embeds_flat.view(batch_size, seq_len, hidden_size)
+
+        outputs = self.language_model(
+            attention_mask=attention_mask,
+            position_ids=position_ids,
+            past_key_values=past_key_values,
+            inputs_embeds=inputs_embeds,
+            use_cache=use_cache,
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+            cache_position=cache_position,
+            num_logits_to_keep=num_logits_to_keep,
+        )
+
+        logits = outputs[0]
+
+        loss = None
+        if labels is not None:
+            # Shift so that tokens < n predict n
+            if attention_mask is not None:
+                shift_attention_mask = attention_mask[..., 1:]
+                shift_logits = logits[..., :-1, :][shift_attention_mask.to(logits.device) != 0].contiguous()
+                shift_labels = labels[..., 1:][shift_attention_mask.to(labels.device) != 0].contiguous()
+            else:
+                shift_logits = logits[..., :-1, :].contiguous()
+                shift_labels = labels[..., 1:].contiguous()
+            # Flatten the tokens
+            loss_fct = nn.CrossEntropyLoss()
+            loss = loss_fct(
+                shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1).to(shift_logits.device)
```

Comment:
this is no longer the default way to compute the lossL https://github.com/huggingface/transformers/blob/f14637a7b540ac1a1b21f6ea919c80a51e287dc6/src/transformers/models/llama/modeling_llama.py#L1129